pymeasure/instruments/velleman/velleman_k8090.py

```diff
@@ -0,0 +1,266 @@
+
+        After switching this waits for a reply from the device. This is only send when
+        a relay actually toggles, otherwise expect a blocking time equal to the
+        communication timeout
+        If speed is important, avoid calling `switch_` unnecessarily.
+        """,
+        set_process=_parse_channels,
+        check_set_errors=True,
+    )
+
+    switch_off = Instrument.setting(
+        "0x12,%s",
+        """
+        Switch off a set of channels. See :attr:`switch_on` for more details.
+        """,
+        set_process=_parse_channels,
+        check_set_errors=True,
+    )
+
+    id = None  # No identification available
+
+    def _make_checksum(self, command, mask, param1, param2):
+        # The formula from the sheet requires twos-complement negation,
+        # this works
+        return 1 + 0xFF - ((self.BYTE_STX + command + mask + param1 + param2) & 0xFF)
+
+    def write(self, command, **kwargs):
+        """The write command specifically for the protocol of the K8090.
+
+        This overrides the method from the ``Instrument`` class.
+
+        Each packet to the device is 7 bytes:
+
+        STX (0x04) - CMD - MASK - PARAM1 - PARAM2 - CHK - ETX (0x0F)
+
+        Where `CHK` is checksum of the package.
+
+        :param command: String like "CMD[, MASK, PARAM1, PARAM2]" - only CMD is mandatory
+        :type command: str
+        """
+
+        # The device can give status updates when we don't expect it,
+        # drop anything from the buffer first
+        if isinstance(self.adapter, VISAAdapter):
+            self.adapter.flush_read_buffer()
+        elif isinstance(self.adapter, SerialAdapter):
+            # The SerialAdapter does not have `flush_read_buffer` implemented
+            self.adapter.connection.flush()
+
+        items_str = command.split(",")
+
+        items = [int(it, 16) for it in items_str]
+
+        cmd = items[0]
+        mask = items[1] if len(items) > 1 else 0
+        param1 = items[2] if len(items) > 2 else 0
+        param2 = items[3] if len(items) > 3 else 0
+
+        checksum = self._make_checksum(cmd, mask, param1, param2)
+
+        content = [
+            self.BYTE_STX,
+            cmd,
+            mask,
+            param1,
+            param2,
+            checksum,
+            self.BYTE_ETX,
+        ]
+
+        self.write_bytes(bytes(content))
+
+    def read(self, **kwargs):
+        """The read command specifically for the protocol of the K8090.
+
+        This overrides the method from the ``instrument`` class.
+
+        See :meth:`write`, replies from the machine use the same format.
+
+        A read will return a list of CMD, MASK, PARAM1 and PARAM2.
+        """
+        # Read all bytes from buffer - until the termchar or a timeout
+        response = self.read_bytes(-1, break_on_termchar=True)
+
+        if len(response) < 7:
+            raise ConnectionError(f"Incoming packet was {len(response)} bytes instead of 7")
```

Comment:
Good idea with that `break_on_termchar`.

Only one question: could the other fields (CMD, MASK, PARAM1/2) have by chance the same value as ETX?

Reply:
Omg you are right, they well can be. The ETX byte is 0X0F (0b00001111 in bits), which is a very reasonable mask for e.g. a switch on / off command.

I will indeed add a line to keep reading of the response is too short.

Reply:
I changed my mind. I had the following written:

```
```